rm/rm-node/src/main/java/org/ow2/proactive/resourcemanager/utils/RRDSigarDataStore.java

```diff
@@ -170,41 +174,23 @@ void sample(RrdDb dataBase, long timeInMs) throws IOException {
             String attrName = names[0];
             String objectName = names[1];
 
-            try {
-                Object attrValue = mbs.getAttribute(new ObjectName(objectName), attrName);
+            // Redirect error stream to avoid sigar exceptions flooding
+            // @see https://github.com/ow2-proactive/scheduling/issues/3068
+            StdErrRedirectionHelper stdErrRedirectionHelper = new StdErrRedirectionHelper();
 
-                if (attrValue instanceof CompositeDataSupport && compositeTypes.get(fullName) != null) {
-                    Object val = ((CompositeDataSupport) attrValue).get(compositeTypes.get(fullName));
-                    if (val != null) {
-                        attrValue = val;
-                    }
-                }
-                sample.setValue(dataSource, Double.parseDouble(attrValue.toString()));
-                if (logger.isTraceEnabled()) {
-                    logger.trace(timeInMs / 1000 + " sampling: " + dataSource + " / " + fullName + " " +
-                                 Double.parseDouble(attrValue.toString()));
-                }
-            } catch (NumberFormatException ex) {
-                // do not save non-numeric values
-                String exceptionAsString = StackTraceUtil.getStackTrace(ex).replaceAll("For input string: \".*\"", "");
-                exceptionAsString = fullName + " / " + exceptionAsString;
-
-                if (!exceptionsThrown.contains(exceptionAsString)) {
-                    exceptionsThrown.add(exceptionAsString);
-                    logger.warn("Non numeric value for " + dataSource + " / " + fullName + ": " + ex.getMessage());
-                } else if (logger.isTraceEnabled()) {
-                    logger.trace("Non numeric value for " + dataSource + " / " + fullName + ": " + ex.getMessage());
-                }
-            } catch (Exception e) {
-                String exceptionAsString = StackTraceUtil.getStackTrace(e);
-                if (!exceptionsThrown.contains(exceptionAsString)) {
-                    exceptionsThrown.add(exceptionAsString);
-                    logger.warn("Error while reading attribute " + attrName + " for object " + objectName + ": ", e);
-                } else if (logger.isTraceEnabled()) {
-                    logger.trace("Error while reading attribute " + attrName + " for object " + objectName + ": ", e);
-                }
+            stdErrRedirectionHelper.redirectStdErr();
+
+            readAttribute(timeInMs, sample, dataSource, fullName, attrName, objectName);
+
+            String errorContent = stdErrRedirectionHelper.readAndReset();
+
+            if (errorContent.length() > 0) {
+                logErrorOnce(errorContent,
+                             "Error while reading attribute " + attrName + " for object " + objectName + ": " +
+                                           System.lineSeparator() + errorContent);
             }
         }
```

Comment:
Don't you need to stop the redirection here?

Reply:
the method readAndReset stops the redirection